### PR TITLE
add a 3 minute timeout to package dependencies

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes:  #941
Update FSharp binaries.

Add in a 3 minute timeout for packagemanagement.

Todo:  Support user configurable timeout for package management.

### CODE REVIEW

**Review by colombod — APPROVED** [review]

File: src/Microsoft.DotNet.Interactive/PackageRestoreContext.cs [hunk 1]
```diff
@@ -22,6 +22,9 @@ public class PackageRestoreContext : IDisposable
         private readonly Dictionary<string, ResolvedPackageReference> _resolvedPackageReferences = new Dictionary<string, ResolvedPackageReference>(StringComparer.OrdinalIgnoreCase);
         private readonly HashSet<string> _restoreSources = new HashSet<string>();
         private readonly DependencyProvider _dependencies;
+        // Todo: design a mechanism for the user to specify this value.
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/PackageRestoreContext.cs:25]: Is there an issue open for this?

**Comment by jonsequitur** [comment]
This was my attempt at this the other day. The PR can probably be closed but there might be one or two details worth keeping:

https://github.com/dotnet/interactive/pull/940/files#diff-7c24de179534b7b7914d7b376592645b4fca966c1a4d87d2ea2bda1a6bcbbe84L224-L227

File: src/Microsoft.DotNet.Interactive/PackageRestoreContext.cs (hunk 1 above)
Comment (KevinRansom) [reply src/Microsoft.DotNet.Interactive/PackageRestoreContext.cs:25]: Yes ... #941

**Comment by KevinRansom** [comment]
> This was my attempt at this the other day. The PR can probably be closed but there might be one or two details worth keeping:
> 
> https://github.com/dotnet/interactive/pull/940/files#diff-7c24de179534b7b7914d7b376592645b4fca966c1a4d87d2ea2bda1a6bcbbe84L224-L227

Let's leave it open for a bit..

**Comment by KevinRansom** [comment]
@jonsequitur . check it out ... it passed ... wooot

**Comment by KevinRansom** [comment]
4. failed
````
    Microsoft.DotNet.Interactive.Tests.CompositeKernelTests.Deferred_commands_on_composite_kernel_can_use_directives [FAIL]
      Expected collection {Microsoft.DotNet.Interactive.Events.CommandSucceeded, Microsoft.DotNet.Interactive.Events.CodeSubmissionReceived, Microsoft.DotNet.Interactive.Events.CompleteCodeSubmissionReceived, Microsoft.DotNet.Interactive.Events.DiagnosticsProduced, Microsoft.DotNet.Interactive.Events.CommandFailed} to contain items {Microsoft.DotNet.Interactive.Events.CodeSubmissionReceived, Microsoft.DotNet.Interactive.Events.CompleteCodeSubmissionReceived, Microsoft.DotNet.Interactive.Events.CommandSucceeded} in order, but Microsoft.DotNet.Interactive.Events.CommandSucceeded (index 2) did not appear (in the right order).
      Stack Trace:
           at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
           at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
           at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
           at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
           at FluentAssertions.Collections.CollectionAssertions`2.ContainInOrder(IEnumerable expected, String because, Object[] becauseArgs)
           at FluentAssertions.Collections.CollectionAssertions`2.ContainInOrder(Object[] expected)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\CompositeKernelTests.cs(457,0): at Microsoft.DotNet.Interactive.Tests.CompositeKernelTests.Deferred_commands_on_composite_kernel_can_use_directives()
        --- End of stack trace from previous location ---
  Finished:    Microsoft.DotNet.Interactive.Tests

````

**Comment by KevinRansom** [comment]
Failed:  
interactive-ci (Build and Test Windows_NT)
````
    Microsoft.DotNet.Interactive.Tests.LanguageServices.CompletionTests.Subsequent_completion_commands_produce_the_expected_results [FAIL]
      Expected collection to contain a single item matching (e Is CompletionsProduced), but no such item was found.
      Stack Trace:
           at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
           at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
           at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
           at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
           at FluentAssertions.Collections.SelfReferencingCollectionAssertions`2.ContainSingle(Expression`1 predicate, String because, Object[] becauseArgs)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\Utility\TestUtility.cs(189,0): at Microsoft.DotNet.Interactive.Tests.Utility.AssertionExtensions.ContainSingle[T](GenericCollectionAssertions`1 should, Func`2 where)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\LanguageServices\CompletionTests.cs(199,0): at Microsoft.DotNet.Interactive.Tests.LanguageServices.CompletionTests.Subsequent_completion_commands_produce_the_expected_results()
        --- End of stack trace from previous location ---
      Output:
        2020-12-21T04:23:10.1471594Z [00-88934786a23547468951f01efc1add7a-d5e6b63daa956a4f-00] [CompletionTests.Subsequent_completion_commands_produce_the_expected_results]  ?  
        2020-12-21T04:23:17.4630643Z [00-88934786a23547468951f01efc1add7a-d5e6b63daa956a4f-00]  ? csharp: CodeSubmissionReceived: using static Microsoft.DotNet.Interactive.Kernel; 
        2020-12-21T04:23:17.5006528Z [00-88934786a23547468951f01efc1add7a-d5e6b63daa956a4f-00]  ? csharp: CompleteCodeSubmissionReceived: using static Microsoft.DotNet.Interactive.Kernel; 
        2020-12-21T04:23:17.5900159Z [00-88934786a23547468951f01efc1add7a-d5e6b63daa956a4f-00]  ? csharp: DiagnosticsProduced 
        2020-12-21T04:23:17.5920783Z [00-88934786a23547468951f01efc1add7a-d5e6b63daa956a4f-00]  ? csharp: CommandFailed: System.OutOfMemoryException: Insufficient memory to continue the execution of the program.
           at System.Runtime.InteropServices.Marshal.AllocHGlobal(IntPtr cb)
           at System.Reflection.Internal.StreamMemoryBlockProvider.ReadMemoryBlockNoLock(Stream stream, Boolean isFileStream, Int64 start, Int32 size)
           at System.Reflection.PortableExecutable.PEReader..ctor(Stream peStream, PEStreamOptions options, Int32 size)
           at Microsoft.CodeAnalysis.ModuleMetadata.CreateFromStream(Stream peStream, PEStreamOptions options)
           at Microsoft.CodeAnalysis.MetadataReference.CreateFromFile(String path, MetadataReferenceProperties properties, DocumentationProvider documentation)
           at Microsoft.CodeAnalysis.Scripting.Hosting.RuntimeMetadataReferenceResolver.ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
           at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
           at Microsoft.CodeAnalysis.CommonReferenceManager`2.TryResolveMissingReference(MetadataReference requestingReference, AssemblyIdentity referenceIdentity, ImmutableDictionary`2& implicitReferenceResolutions, MetadataReferenceResolver resolver, DiagnosticBag resolutionDiagnostics, AssemblyIdentity& resolvedAssemblyIdentity, AssemblyMetadata& resolvedAssemblyMetadata, PortableExecutableReference& resolvedReference)
           at Microsoft.CodeAnalysis.CommonReferenceManager`2.ResolveAndBindMissingAssemblies(TCompilation compilation, ImmutableArray`1 explicitAssemblies, ImmutableArray`1 explicitModules, ImmutableArray`1 explicitReferences, ImmutableArray`1 explicitReferenceMap, MetadataReferenceResolver resolver, MetadataImportOptions importOptions, Boolean supersedeLowerVersions, ArrayBuilder`1 referenceBindings, Dictionary`2 assemblyReferencesBySimpleName, ImmutableArray`1& allAssemblies, ImmutableArray`1& metadataReferences, ImmutableArray`1& resolvedReferences, ImmutableDictionary`2& implicitReferenceResolutions, DiagnosticBag resolutionDiagnostics)
           at Microsoft.CodeAnalysis.CommonReferenceManager`2.Bind(TCompilation compilation, ImmutableArray`1 explicitAssemblies, ImmutableArray`1 explicitModules, ImmutableArray`1 explicitReferences, ImmutableArray`1 explicitReferenceMap, MetadataReferenceResolver resolverOpt, MetadataImportOptions importOptions, Boolean supersedeLowerVersions, Dictionary`2 assemblyReferencesBySimpleName, ImmutableArray`1& allAssemblies, ImmutableArray`1& implicitlyResolvedReferences, ImmutableArray`1& implicitlyResolvedReferenceMap, ImmutableDictionary`2& implicitReferenceResolutions, DiagnosticBag resolutionDiagnostics, Boolean& hasCircularReference, Int32& corLibraryIndex)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.ReferenceManager.CreateAndSetSourceAssemblyFullBind(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.ReferenceManager.CreateSourceAssemblyForCompilation(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.get_SourceModule()
           at Microsoft.CodeAnalysis.CSharp.Symbols.SimpleProgramNamedTypeSymbol.GetSimpleProgramNamedTypeSymbol(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.Symbols.SimpleProgramNamedTypeSymbol.GetSimpleProgramEntryPoint(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetEntryPointAndDiagnostics(CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetEntryPoint(CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.Build[T](Compilation compilation, DiagnosticBag diagnostics, Boolean emitDebugInformation, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.CreateExecutor[T](ScriptCompiler compiler, Compilation compilation, Boolean emitDebugInformation, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.Script`1.GetExecutor(CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.Script`1.RunAsync(Object globals, Func`2 catchException, CancellationToken cancellationToken)
           at Microsoft.DotNet.Interactive.CSharp.CSharpKernel.RunAsync(String code, CancellationToken cancellationToken, Func`2 catchException) in D:\a\1\s\src\Microsoft.DotNet.Interactive.CSharp\CSharpKernel.cs:line 314
           at Microsoft.DotNet.Interactive.CSharp.CSharpKernel.HandleAsync(SubmitCode submitCode, KernelInvocationContext context) in D:\a\1\s\src\Microsoft.DotNet.Interactive.CSharp\CSharpKernel.cs:line 276 
````

**Comment by KevinRansom** [comment]
````
    Microsoft.DotNet.Interactive.Tests.LanguageKernelExtensionLoadingTests.It_loads_extensions_in_specified_directory_via_a_command(language: FSharp) [FAIL]
      System.TimeoutException : The operation has timed out.
      Stack Trace:
        D:\a\1\s\src\Microsoft.DotNet.Interactive\Utility\CommandLine.cs(100,0): at Microsoft.DotNet.Interactive.Utility.CommandLine.Timeout[T](Task`1 source, TimeSpan timeout)
        D:\a\1\s\src\Microsoft.DotNet.Interactive\Utility\CommandLine.cs(50,0): at Microsoft.DotNet.Interactive.Utility.CommandLine.Execute(String command, String args, DirectoryInfo workingDir, Nullable`1 timeout)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\Utility\KernelExtensionTestHelper.cs(167,0): at Microsoft.DotNet.Interactive.Tests.Utility.KernelExtensionTestHelper.CreateExtensionProjectAndBuild(DirectoryInfo projectDir, String code, String extensionName)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\Utility\KernelExtensionTestHelper.cs(103,0): at Microsoft.DotNet.Interactive.Tests.Utility.KernelExtensionTestHelper.CreateExtensionAssembly(DirectoryInfo projectDir, String code, DirectoryInfo copyDllTo, String testName)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\LanguageKernelExtensionLoadingTests.cs(43,0): at Microsoft.DotNet.Interactive.Tests.LanguageKernelExtensionLoadingTests.It_loads_extensions_in_specified_directory_via_a_command(Language language)
        --- End of stack trace from previous location ---
      Output:
        2020-12-23T23:13:35.4418480Z [00-caae2692733e8f4db74d5f0174e46c4d-8ae152d5f7f40740-00] [LanguageKernelExtensionLoadingTests.It_loads_extensions_in_specified_directory_via_a_command]  ?  
        2020-12-23T23:13:35.4669666Z [00-caae2692733e8f4db74d5f0174e46c4d-4a276353b8061b40-00] [CommandLine] [Execute]  ? > C:\hostedtoolcache\windows\dotnet\dotnet.exe build 
        2020-12-23T23:13:35.4670224Z [00-caae2692733e8f4db74d5f0174e46c4d-0e826d7bbd3c8c41-00] [CommandLine] [StartProcess]  ?  
        2020-12-23T23:13:35.4670394Z [00-caae2692733e8f4db74d5f0174e46c4d-0e826d7bbd3c8c41-00] [CommandLine] [StartProcess]  ? (0.0191ms) > C:\hostedtoolcache\windows\dotnet\dotnet.exe build 
        2020-12-23T23:13:35.4688141Z [00-caae2692733e8f4db74d5f0174e46c4d-0e826d7bbd3c8c41-00] [CommandLine] [StartProcess]  ? (1.7948ms)  
        2020-12-23T23:14:59.6692816Z [00-caae2692733e8f4db74d5f0174e46c4d-4a276353b8061b40-00] [CommandLine] [Execute]  ? -> ? (84202.3286ms)  
        2020-12-23T23:14:59.9058831Z [00-caae2692733e8f4db74d5f0174e46c4d-8ae152d5f7f40740-00] [LanguageKernelExtensionLoadingTests.It_loads_extensions_in_specified_directory_via_a_command]  ? (84464.0484ms)  
    Microsoft.DotNet.Interactive.Tests.Utility.AsyncContextTests.async_context_is_available_in_csharp_scripting2 [FAIL]
      System.OutOfMemoryException : Insufficient memory to continue the execution of the program.
      Stack Trace:
           at System.Runtime.InteropServices.Marshal.AllocHGlobal(IntPtr cb)
           at System.Reflection.Internal.StreamMemoryBlockProvider.ReadMemoryBlockNoLock(Stream stream, Boolean isFileStream, Int64 start, Int32 size)
           at System.Reflection.PortableExecutable.PEReader..ctor(Stream peStream, PEStreamOptions options, Int32 size)
           at Microsoft.CodeAnalysis.ModuleMetadata.CreateFromStream(Stream peStream, PEStreamOptions options)
           at Microsoft.CodeAnalysis.MetadataReference.CreateFromFile(String path, MetadataReferenceProperties properties, DocumentationProvider documentation)
           at Microsoft.CodeAnalysis.Scripting.Hosting.RuntimeMetadataReferenceResolver.ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
           at Microsoft.CodeAnalysis.CommonReferenceManager`2.TryResolveMissingReference(MetadataReference requestingReference, AssemblyIdentity referenceIdentity, ImmutableDictionary`2& implicitReferenceResolutions, MetadataReferenceResolver resolver, DiagnosticBag resolutionDiagnostics, AssemblyIdentity& resolvedAssemblyIdentity, AssemblyMetadata& resolvedAssemblyMetadata, PortableExecutableReference& resolvedReference)
           at Microsoft.CodeAnalysis.CommonReferenceManager`2.ResolveAndBindMissingAssemblies(TCompilation compilation, ImmutableArray`1 explicitAssemblies, ImmutableArray`1 explicitModules, ImmutableArray`1 explicitReferences, ImmutableArray`1 explicitReferenceMap, MetadataReferenceResolver resolver, MetadataImportOptions importOptions, Boolean supersedeLowerVersions, ArrayBuilder`1 referenceBindings, Dictionary`2 assemblyReferencesBySimpleName, ImmutableArray`1& allAssemblies, ImmutableArray`1& metadataReferences, ImmutableArray`1& resolvedReferences, ImmutableDictionary`2& implicitReferenceResolutions, DiagnosticBag resolutionDiagnostics)
           at Microsoft.CodeAnalysis.CommonReferenceManager`2.Bind(TCompilation compilation, ImmutableArray`1 explicitAssemblies, ImmutableArray`1 explicitModules, ImmutableArray`1 explicitReferences, ImmutableArray`1 explicitReferenceMap, MetadataReferenceResolver resolverOpt, MetadataImportOptions importOptions, Boolean supersedeLowerVersions, Dictionary`2 assemblyReferencesBySimpleName, ImmutableArray`1& allAssemblies, ImmutableArray`1& implicitlyResolvedReferences, ImmutableArray`1& implicitlyResolvedReferenceMap, ImmutableDictionary`2& implicitReferenceResolutions, DiagnosticBag resolutionDiagnostics, Boolean& hasCircularReference, Int32& corLibraryIndex)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.ReferenceManager.CreateAndSetSourceAssemblyFullBind(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.ReferenceManager.CreateSourceAssemblyForCompilation(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.get_SourceModule()
           at Microsoft.CodeAnalysis.CSharp.Symbols.SimpleProgramNamedTypeSymbol.GetSimpleProgramNamedTypeSymbol(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.Symbols.SimpleProgramNamedTypeSymbol.GetSimpleProgramEntryPoint(CSharpCompilation compilation)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetEntryPointAndDiagnostics(CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetEntryPoint(CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.Build[T](Compilation compilation, DiagnosticBag diagnostics, Boolean emitDebugInformation, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.ScriptBuilder.CreateExecutor[T](ScriptCompiler compiler, Compilation compilation, Boolean emitDebugInformation, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.Script`1.GetExecutor(CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.Script`1.RunAsync(Object globals, Func`2 catchException, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.Scripting.Script`1.RunAsync(Object globals, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.CSharp.Scripting.CSharpScript.RunAsync[T](String code, ScriptOptions options, Object globals, Type globalsType, CancellationToken cancellationToken)
           at Microsoft.CodeAnalysis.CSharp.Scripting.CSharpScript.RunAsync(String code, ScriptOptions options, Object globals, Type globalsType, CancellationToken cancellationToken)
        D:\a\1\s\src\Microsoft.DotNet.Interactive.Tests\Utility\AsyncContextTests.cs(27,0): at Microsoft.DotNet.Interactive.Tests.Utility.AsyncContextTests.async_context_is_available_in_csharp_scripting2()

````

**Comment by KevinRansom** [comment]
It's been running 4 hours so I expect it's hung.

**Comment by KevinRansom** [comment]
````
xUnit.net Console Runner v2.4.1 (64-bit .NET 5.0.0)
  Discovering: Microsoft.DotNet.Interactive.Tests
  Discovered:  Microsoft.DotNet.Interactive.Tests
  Starting:    Microsoft.DotNet.Interactive.Tests
    Microsoft.DotNet.Interactive.Tests.VariableSharingTests.Directives_can_access_local_kernel_variables [SKIP]
      not implemented
    Microsoft.DotNet.Interactive.Tests.VariableSharingTests.Internal_types_are_shared_as_their_most_public_supertype [SKIP]
      WIP
    Microsoft.DotNet.Interactive.Tests.LanguageKernelTests.diagnostics_can_be_produced_from_multiple_subkernels [SKIP]
      The first failed sub-command cancels all subsequent command executions; the second kernel doesn't get a chance to report.
  Finished:    Microsoft.DotNet.Interactive.Tests
=== TEST EXECUTION SUMMARY ===
   Microsoft.DotNet.Interactive.Tests  Total: 652, Errors: 0, Failed: 1, Skipped: 3, Time: 828.171s
=== COMMAND LINE ===
"C:\hostedtoolcache\windows\dotnet\dotnet.exe" exec --depsfile "D:\a\1\s\artifacts\bin\Microsoft.DotNet.Interactive.Tests\Release\net5.0\Microsoft.DotNet.Interactive.Tests.deps.json" --runtimeconfig "D:\a\1\s\artifacts\bin\Microsoft.DotNet.Interactive.Tests\Release\net5.0\Microsoft.DotNet.Interactive.Tests.runtimeconfig.json"  "D:\a\1\s\.packages\xunit.runner.console/2.4.1/tools/netcoreapp2.0/xunit.console.dll" "D:\a\1\s\artifacts\bin\Microsoft.DotNet.Interactive.Tests\Release\net5.0\Microsoft.DotNet.Interactive.Tests.dll" -noautoreporters -xml "D:\a\1\s\artifacts\TestResults\Release\Microsoft.DotNet.Interactive.Tests_net5.0_x64.xml" -html "D:\a\1\s\artifacts\TestResults\Release\Microsoft.DotNet.Interactive.Tests_net5.0_x64.html" -parallel none > "D:\a\1\s\artifacts\log\Release\Microsoft.DotNet.Interactive.Tests_net5.0_x64.log" 2>&1

````

**Comment by KevinRansom** [comment]
failed

**Comment by KevinRansom** [comment]
Microsoft.DotNet.Interactive.Tests.KernelInvocationContextTests.When_a_command_spawns_another_command_then_parent_context_is_not_complete_until_child_context_is_complete [FAIL]

**Comment by KevinRansom** [comment]
passed

**Pull request closed without merging**